dan_layer/consensus/src/hotstuff/on_propose.rs

```diff
@@ -64,26 +61,55 @@ where
 
         // The scope here is due to a shortcoming of rust. The tx is dropped at tx.commit() but it still complains that
         // the non-Send tx could be used after the await point, which is not possible.
-        let (involved_shards, next_block) = {
+        let involved_shards;
+        let next_block;
+        {
             let mut tx = self.store.create_write_tx()?;
             let high_qc = HighQc::get(&mut *tx, epoch)?;
             let high_qc = high_qc.get_quorum_certificate(&mut *tx)?;
 
             let parent_block = leaf_block.get_block(&mut *tx)?;
 
-            let next_block = self.build_next_block(&mut tx, epoch, &parent_block, high_qc, validator_id)?;
-            let involved_shards = next_block.find_involved_shards(&mut *tx)?;
+            // Edge case: on_beat is called again before the leaf has changed - so check that we haven't already
+            // proposed for this leaf
+            if let Some(child) = parent_block.get_child(&mut *tx).optional()? {
+                debug!(
+                    target: LOG_TARGET,
+                    "🌿 Already proposed block for current leaf {} ({}) - child {} ({})",
+                    parent_block.id(),
+                    parent_block.height(),
+                    child.id(),
+                    child.height(),
+                );
+                tx.rollback()?;
+                return Ok(());
+            }
+
+            next_block = self.build_next_block(&mut *tx, epoch, &parent_block, high_qc, validator_id)?;
+            if next_block.exists(&mut *tx)? {
```

Comment:
this check is probably not necessary since it should be caught by the above statement, but probably doesn't hurt

Reply:
yeah you're right, i put that in at first but then thought of the above check after and forgot to remove this one. will leave it in for now, as you say, doesnt hurt